src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,181 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  //     - MulNode(a, CON)
+  // This optimization technically only produces MulNode(CON, a), but we might as well transform (a * CON) + a, too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // swap ConNode to lhs for easier matching
+    if (!con->is_Con()) {
+      swap(con, base);
+    }
+
+    if (con->is_top()) {
+      return nullptr;
+    }
+
+    *multiplier = con->get_integer_as_long(bt);
+    return base;
+  }
+
+  return nullptr;
+}
+
+// Match `(a << CON1) + (a << CON2)`, extract `a` and `(1 << CON1) + (1 << CON2)`
+// Note that one of the term of the addition could simply be `a` (i.e., a << 0)
+Node* AddNode::find_power_of_two_addition_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // Look for patterns:
+  //     - AddNode(LShiftNode(a, CON), LShiftNode(a, CON)/a)
+  //     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+  // given that lhs is different from rhs
+  if (n->Opcode() == Op_Add(bt) && n->in(1) != n->in(2)) {
+    Node* lhs = n->in(1);
+    Node* rhs = n->in(2);
+
+    // swap LShiftNode to lhs for easier matching
+    if (lhs->Opcode() != Op_LShift(bt)) {
+      swap(lhs, rhs);
+    }
+
+    // AddNode(LShiftNode(a, CON), *)?
+    if (lhs->Opcode() != Op_LShift(bt) || !lhs->in(2)->is_Con()) {
+      return nullptr;
+    }
+
+    jlong lhs_multiplier = 0;
+    if (multiplier != nullptr) {
+      Node* con = lhs->in(2);
+      if (con->is_top()) {
+        return nullptr;
+      }
+
+      lhs_multiplier = (jlong) 1 << con->get_int();
+    }
+
+    // AddNode(LShiftNode(a, CON), a)?
+    if (lhs->in(1) == rhs) {
+      if (multiplier != nullptr) {
+        *multiplier = lhs_multiplier + 1;
+      }
+
+      return rhs;
+    }
+
+    // AddNode(LShiftNode(a, CON), LShiftNode(a, CON2))?
+    if (rhs->Opcode() == Op_LShift(bt) && lhs->in(1) == rhs->in(1) && rhs->in(2)->is_Con()) {
+      if (multiplier != nullptr) {
+        Node* con = rhs->in(2);
+        if (con->is_top()) {
+          return nullptr;
+        }
+
+        *multiplier = lhs_multiplier + ((jlong) 1 << con->get_int());
+      }
+
+      return lhs->in(1);
+    }
+
+    return nullptr;
+  }
+
+  return nullptr;
+}
+
+// Match `(a << CON) - a`, extract `a` and `(1 << CON) - 1`
+Node* AddNode::find_power_of_two_subtraction_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // Look for pattern: SubNode(LShiftNode(a, CON), a)
+  if (n->Opcode() == Op_Sub(bt) && n->in(1)->Opcode() == Op_LShift(bt) && n->in(1)->in(2)->is_Con()) {
+    Node* lshift = n->in(1);
+    Node* base = n->in(2);
+
+    Node* lshift_base = lshift->in(1);
+    if (lshift_base != base) {
+      return nullptr;
+    }
+
+    Node* con = lshift->in(2);
+    if (con->is_top()) {
+      return nullptr;
+    }
+
+    // We can't simply return the lshift node even if ((a << CON) - a) + a cancels out. Ideal() must return a new node.
+    *multiplier = ((jlong) 1 << con->get_int()) - 1;
```

Comment:
Can't this be an `Identity()` transformation where you can return existing nodes?

Reply:
That's great!